kymatio/scattering2d/tests/test_tensorflow_scattering2d.py

```diff
@@ -0,0 +1,366 @@
+
+        y = modulus(x)
+        u = np.squeeze(np.sqrt(np.real(x) ** 2 + np.imag(x) ** 2))
+        v = y
+        assert np.allclose(u, v)
+
+class TestSubsampleFourier:
+    @pytest.mark.parametrize('backend', backends)
+    def test_SubsampleFourier(self, backend):
+        subsample_fourier = backend.subsample_fourier
+
+        x = (np.random.rand(100, 128, 128)
+             + 1J * np.random.rand(100, 128, 128))
+
+        y = np.zeros((100, 8, 8), dtype=np.complex128)
+
+        from itertools import product
+        for i, j in product(range(8), range(8)):
+            for m, n in product(range(16), range(16)):
+                y[..., i, j] += x[..., i + m * 8, j + n * 8]
+
+        y /= 16 ** 2
+
+        z = subsample_fourier(x, k=16)
+        assert np.allclose(y, z)
+
+class TestCDGMM:
+    @pytest.fixture(params=(False, True))
+    def data(self, request):
+        real_filter = request.param
+        x = (np.random.randn(100, 128, 128)
+             + 1J * np.random.randn(100, 128, 128))
+        filt = (np.random.randn(128, 128)
+                + 1J * np.random.randn(128, 128))
+        y = (np.random.randn(100, 128, 128)
+             + 1J * np.random.randn(100, 128, 128))
+
+        if real_filter:
+            filt = np.real(filt)
+
+        y = x * filt
+
+        return x, filt, y
+
+    @pytest.mark.parametrize('backend', backends)
+    @pytest.mark.parametrize('inplace', (False, True))
+    def test_cdgmm_forward(self, data, backend, inplace):
+        x, filt, y = data
+
+        z = backend.cdgmm(x, filt, inplace=inplace)
+
+        assert np.allclose(y, z)
+
+    @pytest.mark.parametrize('backend', backends)
+    def test_cdgmm_exceptions(self, backend):
+        with pytest.raises(TypeError) as record:
+            backend.cdgmm(np.empty((3, 4, 5)).astype(np.float64),
+                          np.empty((4, 5)).astype(np.complex128))
+        assert 'first input must be complex' in record.value.args[0]
+
+        with pytest.raises(RuntimeError) as record:
+            backend.cdgmm(np.empty((3, 4, 5)).astype(np.complex128),
+                          np.empty((3, 4, 5)).astype(np.complex128))
+        assert 'second input must be 2' in record.value.args[0]
+
+        with pytest.raises(TypeError) as record:
+            backend.cdgmm(np.empty((3, 4, 5)).astype(np.complex128),
+                          np.empty((4, 5)).astype(np.int64))
+        assert 'second input must be complex or real' in record.value.args[0]
+
+        with pytest.raises(RuntimeError) as record:
+            backend.cdgmm(np.empty((3, 4, 5)).astype(np.complex128),
+                          np.empty((4, 6)).astype(np.complex128))
+        assert 'not compatible for multiplication' in record.value.args[0]
+
+
+
+class TestFFT:
+    @pytest.mark.parametrize('backend', backends)
+    def test_fft(self, backend):
+        x = np.random.randn(2, 2) + 1J * np.random.randn(2, 2)
+        x = x.astype('complex64')
```

Comment:
Here is the difference! we have to cast to `complex64` in order for the test here to pass. On the Numpy, we don't need to cast to `complex64`. 

This is the only difference, else most things are the same. We should probably look into unifying both set of tests.

Reply:
isn't it a sort of cheat? I'm not sure how tf works

Reply:
Explain what you mean by cheat? 

The casting is needed in order for the tests to pass.

Reply:
Cheat? No I don't think so. Both Torch and TF only allow multiplication (and addition, etc.) of arrays that have the same dtype. NP does not, so we get away with certain things. If you look at the corresponding Torch test, it also has it in single precision (although it's not explicit since it's the default precision for random numbers in Torch).

Reply:
ok.